File: src/main/java/uk/co/squadlist/web/urls/UrlBuilder.java
package uk.co.squadlist.web.urls;

import com.google.common.base.Joiner;
import com.google.common.collect.Lists;
import org.apache.commons.codec.digest.DigestUtils;
import org.apache.http.client.utils.URIBuilder;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import uk.co.squadlist.web.context.InstanceConfig;
import uk.co.squadlist.web.localisation.GoverningBody;
import uk.co.squadlist.web.model.*;

import java.net.URISyntaxException;
import java.util.List;

@Component("urlBuilder")
public class UrlBuilder {

	private final String baseUrl;
	private final InstanceConfig instanceConfig;
	private final SeoLinkBuilder seoLinkBuilder;
	private String apiUrl;

	@Autowired
	public UrlBuilder(@Value("${baseUrl}") String baseUrl, InstanceConfig instanceConfig, SeoLinkBuilder seoLinkBuilder, @Value("${apiUrl}") String apiUrl) {
		this.baseUrl = baseUrl;
		this.instanceConfig = instanceConfig;
		this.seoLinkBuilder = seoLinkBuilder;
		this.apiUrl = apiUrl;
	}

	public String applicationUrl(String uri) {
		return getBaseUrl() + uri;
	}

	public String loginUrl() {
		return applicationUrl("/login");
	}

	public String staticUrl(String uri) {
		return applicationUrl("/static/") + uri;
	}

	public String boatUrl(Boat boat) {
		return applicationUrl("/boats/" + boat.getId());
	}
	public String memberUrl(Member member) {
		return memberUrl(member.getId());
	}

	public String memberResetPasswordUrl(Member member) {
		return memberUrl(member.getId()) + "/reset";
	}

	public String memberUrl(String memberId) {
		return applicationUrl("/member/" + memberId);
	}

	public String makeActive(Member member) {
		return memberUrl(member.getId()) + "/make-active";
	}

	public String makeInactive(Member member) {
		return memberUrl(member.getId()) + "/make-inactive";
	}

	public String delete(Member member) {
		return memberUrl(member.getId()) + "/delete";
	}

	public String delete(Squad squad) {
		return applicationUrl("/squad/" + squad.getId() + "/delete");
	}

	public String newMemberUrl() {
		return applicationUrl("/member/new");
	}

	public String newSquadUrl() {
		return applicationUrl("/squad/new");
	}

	public String editSquadUrl(Squad squad) {
		return applicationUrl("/squad/" + squad.getId() + "/edit");
	}

	public String adminUrl() {
		return applicationUrl("/admin");
	}

	public String outingsUrl() {
		return applicationUrl("/outings");
	}

	public String outingUrl(Outing outing) {
		return outingsUrl() + "/" + outing.getId();
	}

	public String outingAvailabilityCsv(Outing outing) {
		return outingUrl(outing) + ".csv";
	}

	public String outingCloseUrl(Outing outing) {
		return outingUrl(outing) + "/close";
	}

	public String deleteOuting(Outing outing) {
		return outingUrl(outing) + "/delete";
	}

	public String outingReopenUrl(Outing outing) {
		return outingUrl(outing) + "/reopen";
	}

	public String outingEditUrl(Outing outing) {
		return outingUrl(outing) + "/edit";
	}

	public String outings(Squad squad) {
		return applicationUrl("/outings?squad=" + squad.getId());
	}

	public String outings(Squad squad, String month) {
		return outings(squad) + "&month=" + month;
	}

	public String availability(Squad squad) {
		return applicationUrl("/availability/" + squad.getId());
	}

	public String availability(Squad squad, String month) {
		return availability(squad) + "?month=" + month;
	}

	public String editInstanceSettings() {
		return adminUrl() + "/instance";
	}

	public String editMemberUrl(Member member) {
		return memberUrl(member) + "/edit";
	}

	public String editMemberUrl(String memberId) {
		return memberUrl(memberId) + "/edit";
	}

	public String entryDetailsCsv(Squad squad) {
		return applicationUrl("/entrydetails/" + squad.getId() + ".csv");
	}

	public String entryDetailsCsv(List<Member> members) {
		try {
			final URIBuilder url = new URIBuilder(applicationUrl("/entrydetails/selected.csv"));

			final List<String> memberIds = Lists.newArrayList();
			for (Member member : members) {
				memberIds.add(member.getId());
			}
			url.addParameter("members", Joiner.on(",").join(memberIds));

			return url.toString();

		} catch (URISyntaxException e) {
			throw new RuntimeException(e);
		}
	}

	public String changePassword() {
		return applicationUrl("/change-password");
	}

	public String socialMediaAccounts() {
		return applicationUrl("/social");
	}

	public String linkFacebookUrl() {
		return applicationUrl("/social/facebook/link");
	}

	public String removeFacebookUrl() {
		return applicationUrl("/social/facebook/remove");
	}

	public String getLinkFacebookCallbackUrl() {
		return applicationUrl("/social/facebook/link/callback");
	}

	public String facebookSigninCallbackUrl() {
		return applicationUrl("/social/facebook/signin/callback");
	}

	public String facebookSignin() {
		return applicationUrl("/social/facebook/signin");
	}

	public String getBaseUrl() {
		return baseUrl.replace("INSTANCE", instanceConfig.getVhost());
	}

	public String contactsUrl() {
		return applicationUrl("/contacts");
	}
	public String contactsUrl(Squad prefferredSquad) {
		return appendSquad(prefferredSquad, contactsUrl());
	}

	public String outingsUrl(Squad prefferredSquad) throws URISyntaxException {
		final URIBuilder url = new URIBuilder(outingsUrl());
		if (prefferredSquad != null) {
			url.addParameter("squad", prefferredSquad.getId());
		}
		return url.build().toString();
	}

	public String outingsRss(String userid, Instance instance) throws URISyntaxException {
		final URIBuilder url = new URIBuilder(applicationUrl("/rss"));
		url.addParameter("user", userid);
		url.addParameter("key", generateFeedKeyFor(userid, instance));
		return url.build().toString();
	}

	public String outingsIcal(String userid, Instance instance) throws URISyntaxException {
		String webcalBaseUrl = getBaseUrl().replaceAll("^https?://", "webcal://");
		final URIBuilder url = new URIBuilder(webcalBaseUrl  + "/ical");
		url.addParameter("user", userid);
		url.addParameter("key", generateFeedKeyFor(userid, instance));
		return url.build().toString();
	}

	public String availabilityUrl(Squad prefferredSquad) {
		return appendSquad(prefferredSquad, availabilityUrl());
	}

	public String governingBody(GoverningBody governingBody) {
		return applicationUrl("/governing-body/" + seoLinkBuilder.makeSeoLinkFor(governingBody.getName()));
	}

	public String staticImage(String filename) {
		return apiUrl + "/static/" + filename + ".jpg";
	}

	public String newAvailabilityOptionUrl() {
		return applicationUrl("/availability-option/new");
	}

	public String deleteAvailabilityOptionUrl(AvailabilityOption availabilityOption) {
		return applicationUrl("/availability-option/" + availabilityOption.getId() + "/delete");
	}

	public String editAvailabilityOptionUrl(AvailabilityOption availabilityOption) {
		return applicationUrl("/availability-option/" + availabilityOption.getId() + "/edit");
	}

	public String editAdmins() {
		return adminUrl() + "/admins";
	}

	public String adminExportMembersAsCSV() {
		return adminUrl() + "/export/members.csv";
	}

	private String availabilityUrl() {
		return applicationUrl("/availability");
	}

	private String appendSquad(Squad prefferredSquad, final String baseUrl) {
		return prefferredSquad != null ? baseUrl + "/" + prefferredSquad.getId() : baseUrl;
	}

	private String generateFeedKeyFor(String userid, Instance instance) {
		return DigestUtils.md5Hex(instance.getId() + userid);
	}

	public String mailto(List<String> emails) {
		if (emails.isEmpty()) {
			return null;
		}
		return "mailto:" + Joiner.on(",").join(emails);
	}

}

File: src/main/kotlin/uk/co/squadlist/web/urls/UrlBuilder.kt
package uk.co.squadlist.web.urls

import com.google.common.base.Joiner
import com.google.common.collect.Lists
import org.apache.commons.codec.digest.DigestUtils
import org.apache.http.client.utils.URIBuilder
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import uk.co.squadlist.web.context.InstanceConfig
import uk.co.squadlist.web.localisation.GoverningBody
import uk.co.squadlist.web.model.*
import java.net.URISyntaxException

@Component
class UrlBuilder (@param:Value("\${baseUrl}") private val baseUrl: String,
                  val instanceConfig: InstanceConfig,
                  val seoLinkBuilder: SeoLinkBuilder,
                  @param:Value("\${apiUrl}") private val apiUrl: String,
                  @param:Value("\${profilePicturesUrl}") private val profilePicturesUrl: String) {

    val linkFacebookCallbackUrl: String
        get() = applicationUrl("/social/facebook/link/callback")

    fun applicationUrl(uri: String): String {
        return getBaseUrl() + uri
    }

    fun loginUrl(): String {
        return applicationUrl("/login")
    }

    fun resetPassword(): String {
        return applicationUrl("/reset-password")
    }

    fun staticUrl(uri: String): String {
        return applicationUrl("/assets/") + uri
    }

    fun boatUrl(boat: Boat): String {
        return applicationUrl("/boats/" + boat.id)
    }

    fun memberUrl(member: Member): String {
        return memberUrl(member.id)
    }

    fun memberResetPasswordUrl(member: Member): String {
        return memberUrl(member.id) + "/reset"
    }

    fun memberUrl(memberId: String): String {
        return applicationUrl("/member/$memberId")
    }

    fun makeActive(member: Member): String {
        return memberUrl(member.id) + "/make-active"
    }

    fun makeInactive(member: Member): String {
        return memberUrl(member.id) + "/make-inactive"
    }

    fun delete(member: Member): String {
        return memberUrl(member.id) + "/delete"
    }

    fun delete(squad: Squad): String {
        return applicationUrl("/squad/" + squad.id + "/delete")
    }

    fun newMemberUrl(): String {
        return applicationUrl("/member/new")
    }

    fun newSquadUrl(): String {
        return applicationUrl("/squad/new")
    }

    fun editSquadUrl(squad: Squad): String {
        return applicationUrl("/squad/" + squad.id + "/edit")
    }

    fun adminUrl(): String {
        return applicationUrl("/admin")
    }

    fun outingsUrl(): String {
        return applicationUrl("/outings")
    }

    fun outingUrl(outing: Outing): String {
        return outingsUrl() + "/" + outing.id
    }

    fun outingAvailabilityCsv(outing: Outing): String {
        return outingUrl(outing) + ".csv"
    }

    fun outingCloseUrl(outing: Outing): String {
        return outingUrl(outing) + "/close"
    }

    fun deleteOuting(outing: Outing): String {
        return outingUrl(outing) + "/delete"
    }

    fun outingReopenUrl(outing: Outing): String {
        return outingUrl(outing) + "/reopen"
    }

    fun outingEditUrl(outing: Outing): String {
        return outingUrl(outing) + "/edit"
    }

    fun entryDetails(squad: Squad): String {
        return applicationUrl("/entry-details?squad=" + squad.id)
    }

    fun outings(squad: Squad): String {
        return applicationUrl("/outings?squad=" + squad.id)
    }

    fun outings(squad: Squad, month: String): String {
        return outings(squad) + "&month=" + month
    }

    fun availability(squad: Squad): String {
        return applicationUrl("/availability/" + squad.id)
    }

    fun availability(squad: Squad, month: String): String {
        return availability(squad) + "?month=" + month
    }

    fun editInstanceSettings(): String {
        return adminUrl() + "/instance"
    }

    fun editMemberUrl(member: Member): String {
        return memberUrl(member) + "/edit"
    }

    fun editMemberUrl(memberId: String): String {
        return memberUrl(memberId) + "/edit"
    }

    fun entryDetailsCsv(squad: Squad): String {
        return applicationUrl("/entrydetails/" + squad.id + ".csv")
    }

    fun entryDetailsCsv(members: List<Member>): String {
        try {
            val url = URIBuilder(applicationUrl("/entrydetails/selected.csv"))

            val memberIds = Lists.newArrayList<String>()
            for (member in members) {
                memberIds.add(member.id)
            }
            url.addParameter("members", Joiner.on(",").join(memberIds))

            return url.toString()

        } catch (e: URISyntaxException) {
            throw RuntimeException(e)
        }

    }

    fun changePassword(): String {
        return applicationUrl("/change-password")
    }

    fun socialMediaAccounts(): String {
        return applicationUrl("/social")
    }

    fun linkFacebookUrl(): String {
        return applicationUrl("/social/facebook/link")
    }

    fun removeFacebookUrl(): String {
        return applicationUrl("/social/facebook/remove")
    }

    fun facebookSigninCallbackUrl(): String {
        return applicationUrl("/social/facebook/signin/callback")
    }

    fun facebookSignin(): String {
        return applicationUrl("/social/facebook/signin")
    }

    fun getBaseUrl(): String {
        return baseUrl.replace("INSTANCE", instanceConfig.vhost)
    }

    fun contactsUrl(): String {
        return applicationUrl("/contacts")
    }

    fun contactsUrl(prefferredSquad: Squad): String {
        return appendSquad(prefferredSquad, contactsUrl())
    }

    @Throws(URISyntaxException::class)
    fun outingsUrl(prefferredSquad: Squad?): String {
        val url = URIBuilder(outingsUrl())
        if (prefferredSquad != null) {
            url.addParameter("squad", prefferredSquad.id)
        }
        return url.build().toString()
    }

    @Throws(URISyntaxException::class)
    fun outingsRss(userid: String, instance: Instance): String {
        val url = URIBuilder(applicationUrl("/rss"))
        url.addParameter("user", userid)
        url.addParameter("key", generateFeedKeyFor(userid, instance))
        return url.build().toString()
    }

    @Throws(URISyntaxException::class)
    fun outingsIcal(userid: String, instance: Instance): String {
        val webcalBaseUrl = getBaseUrl().replace("^https?://".toRegex(), "webcal://")
        val url = URIBuilder("$webcalBaseUrl/ical")
        url.addParameter("user", userid)
        url.addParameter("key", generateFeedKeyFor(userid, instance))
        return url.build().toString()
    }

    fun availabilityUrl(prefferredSquad: Squad): String {
        return appendSquad(prefferredSquad, availabilityUrl())
    }

    fun governingBody(governingBody: GoverningBody): String {
        return applicationUrl("/governing-body/" + seoLinkBuilder.makeSeoLinkFor(governingBody.name))
    }

    fun staticImage(filename: String): String {
        return "$apiUrl/static/$filename.jpg"
    }

    fun profileImage(filename: String): String {
        return "$profilePicturesUrl/$filename.jpg"    // TODO ideally all of this would be provided by the API
    }

    fun newAvailabilityOptionUrl(): String {
        return applicationUrl("/availability-option/new")
    }

    fun deleteAvailabilityOptionUrl(availabilityOption: AvailabilityOption): String {
        return applicationUrl("/availability-option/" + availabilityOption.id + "/delete")
    }

    fun editAvailabilityOptionUrl(availabilityOption: AvailabilityOption): String {
        return applicationUrl("/availability-option/" + availabilityOption.id + "/edit")
    }

    fun editAdmins(): String {
        return adminUrl() + "/admins"
    }

    fun adminExportMembersAsCSV(): String {
        return adminUrl() + "/export/members.csv"
    }

    private fun availabilityUrl(): String {
        return applicationUrl("/availability")
    }

    private fun appendSquad(prefferredSquad: Squad?, baseUrl: String): String {
        return if (prefferredSquad != null) baseUrl + "/" + prefferredSquad.id else baseUrl
    }

    private fun generateFeedKeyFor(userid: String, instance: Instance): String {
        return DigestUtils.md5Hex(instance.id + userid)
    }

    fun mailto(emails: List<String>): String? {
        return if (emails.isEmpty()) {
            null
        } else "mailto:" + Joiner.on(",").join(emails)
    }

}
